### PR TITLE
BUG: Group-by numeric type-coercion with datetime

### LINKED ISSUE #14423: BUG: unwanted numeric coercion after groupby-apply
xref #14873 (boolean casts)
xref #14849 (datetime)

#### A small, complete example of the issue

``` python
import pandas as pd

def predictions(tool):
    out = pd.Series(index=['p1', 'p2', 'useTime'], dtype=object)
    if 'step1' in list(tool.State):
        out['p1'] = str(tool[tool.State == 'step1'].Machine.values[0])
    if 'step2' in list(tool.State):
        out['p2'] = str(tool[tool.State == 'step2'].Machine.values[0])
        out['useTime'] = str(tool[tool.State == 'step2'].oTime.values[0])
    return out


df1 = pd.DataFrame({'Key': ['B', 'B', 'A', 'A'],
                   'State': ['step1', 'step2', 'step1', 'step2'],
                   'oTime': ['', '2016-09-19 05:24:33', '', '2016-09-19 23:59:04'],
                   'Machine': ['23', '36L', '36R', '36R']})

df2 = df1.copy()
df2.oTime = pd.to_datetime(df2.oTime)


pred1 = df1.groupby('Key').apply(predictions)
pred2 = df2.groupby('Key').apply(predictions)

print(pred1)
print(pred2)
```
#### Actual Output:

```
      p1   p2              useTime
Key                               
A    36R  36R  2016-09-19 23:59:04
B     23  36L  2016-09-19 05:24:33
       p1   p2                        useTime
Key                                          
A     NaN  36R  2016-09-19T23:59:04.000000000
B    23.0  36L  2016-09-19T05:24:33.000000000
```
#### Expected Output

pred1 and pred2 should have the same values in column p1. 
pred1 is correct whereas pred2 is changing type to float64.
#### Output of `pd.show_versions()`

<details>
## INSTALLED VERSIONS

commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Darwin
OS-release: 14.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.19.0
nose: None
pip: 8.1.2
setuptools: 27.2.0
Cython: None
numpy: 1.11.2
scipy: 0.18.1
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: None
patsy: None
dateutil: 2.5.3
pytz: 2016.7
blosc: None
bottleneck: None
tables: 3.2.3.1
numexpr: 2.6.1
matplotlib: 1.5.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.1.0
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
pandas_datareader: None
</details>

### LINKED ISSUE #15421: Unexpected string->float conversion in DataFrame.groupby().apply()
#### Code Sample, a copy-pastable example if possible

```python
# Your code here
import pandas as pd
df = pd.DataFrame({'A':[10, 20, 30], 'B': [ 'foo', '3', '4'], 'T': [pd.Timestamp("12:31:22")]*3})
def get_B(g):
    return g.iloc[0][['B']]
print df.groupby('A').apply(get_B)

# Observed output:
      B
A      
10  NaN
20  3.0
30  4.0
```

#### Problem description
groupby.apply() does an unexpected conversion from string to float for column 'B' in the example above.  The bug is triggered only when both of the following happen:
1.  A column ('B' in the example above) has string values, some of which are parseable as numbers and some which are not.
2. Another column ('T' in the example above) in the dataframe has timestamps.

#### Expected Output
```
      B
A      
10  foo
20  3
30  4
```
#### Output of ``pd.show_versions()``

<details>
# Paste the output here pd.show_versions() here

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.13.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.19.2
nose: 1.3.7
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.11.3
scipy: 0.18.1
statsmodels: 0.6.1
xarray: None
IPython: 5.1.0
sphinx: 1.5.1
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: 1.2.0
tables: 3.2.2
numexpr: 2.6.1
matplotlib: 1.5.1
openpyxl: 2.4.0
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.6
lxml: 3.7.2
bs4: 4.5.3
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.1.4
pymysql: None
psycopg2: None
jinja2: 2.8.1
boto: 2.45.0
pandas_datareader: 0.2.1
</details>

### LINKED ISSUE #15670: Date Type Corrupting Other Types in Group-by/Apply
#### Code Sample, a copy-pastable example if possible

```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'Number' : [1, 2], 'Date' : ["2017-03-02"] * 2, 'Str' : ["foo", "inf"]})

In [3]: df
Out[3]:
         Date  Number  Str
0  2017-03-02       1  foo
1  2017-03-02       2  inf

In [4]: df.groupby(['Number']).apply(lambda x: x.iloc[0])
Out[4]:
              Date  Number  Str
Number
1       2017-03-02       1  foo
2       2017-03-02       2  inf

In [5]: df.Date = pd.to_datetime(df.Date)

In [6]: df
Out[6]:
        Date  Number  Str
0 2017-03-02       1  foo
1 2017-03-02       2  inf

In [7]: df.groupby(['Number']).apply(lambda x: x.iloc[0])
Out[7]:
             Date  Number  Str
Number
1      2017-03-02       1  NaN
2      2017-03-02       2  inf

```

#### Problem description

When I change the type of the Date column to a Pandas datetime, it causes other columns' types to change in unexpected ways when doing a group-by/apply.  Notice the contents of the "Str" column changes to a numeric type in the final group-by/apply (a contributing factor is probably that one of the elements is the string "inf").  The "inf" value has become inf, and the "foo" value has become NaN.

#### Expected Output

I expect the Str column to remain a string type, and contain the original strings.  I.e.:

```python
        Date  Number  Str
0 2017-03-02       1  foo
1 2017-03-02       2  inf
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-327.10.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.1
nose: 1.3.7
pip: None
setuptools: 0.6
Cython: 0.24.1
numpy: 1.11.1
scipy: 0.18.0
statsmodels: 0.6.1
xarray: 0.7.0
IPython: 5.0.0
sphinx: 1.3.5
patsy: 0.4.1
dateutil: 2.5.3
pytz: 2016.6.1
blosc: None
bottleneck: 1.1.0
tables: 3.2.3.1
numexpr: 2.6.1
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.2
lxml: 3.6.1
bs4: 4.4.1
html5lib: 0.999
httplib2: None
apiclient: None
sqlalchemy: 1.0.13
pymysql: 0.6.7.None
psycopg2: 2.5.4 (dt dec pq3 ext)
jinja2: 2.8
boto: 2.40.0
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by gwpdt**
closes #14423
closes #15421
closes #15670

During a group-by/apply on a DataFrame, in the presence of one or more
DateTime-like columns, Pandas would incorrectly coerce the type of all
other columns to numeric.  E.g. a String column would be coerced to
numeric, producing NaNs.

### CODE REVIEW

File: pandas/core/groupby.py [hunk 1]
```diff
@@ -3566,7 +3568,8 @@ def first_non_None_value(values):
                 # as we are stacking can easily have object dtypes here
                 so = self._selected_obj
                 if (so.ndim == 2 and so.dtypes.apply(is_datetimelike).any()):
-                    result = result._convert(numeric=True)
+                    result = result.apply(
+                        lambda x: pd.to_numeric(x, errors='ignore'))
```
Comment (jreback) [inline pandas/core/groupby.py]: don't import pd
import to_numeric from pandas.util.misc (inside the function)

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -4314,6 +4314,16 @@ def test_cummin_cummax(self):
         expected = pd.Series([1, 2, 1], name='b')
         tm.assert_series_equal(result, expected)
 
+    def test_numeric_coercion(self):
+        # GH 14423
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: there are a couple of issues marked duplicate can u add there tests here as well (inside this test is ok)

File: pandas/core/groupby.py (hunk 1 above)
Comment (jreback) [inline pandas/core/groupby.py]: or u can import pd inside the function is ok too

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm couple of comments
pls add a whatsnew note in bug fixes
ping when green or need review

**Comment by jreback** [comment]
pls add tests from these two

xref #14873 (boolean casts)
xref #14849 (datetime)

if this change doesn't fix these, pls lmk.

**Comment by gwpdt** [comment]
I tried adding the test for #14873 - my fix does not resolve this issue.  It's similar, but unrelated.   It seems to be this line in _python_agg_general:

`output[name] = self._try_cast(result, obj, numeric_only=True)`

that seems to favour converting the result of the apply back to the same type as the original series.

**Comment by jreback** [comment]
@gwpdt ok, we can treat that issue independently.

**Comment by gwpdt** [comment]
I also tried adding the test for #14849 - my fix does not resolve this issue either.

The issue with that test is that it is testing for preservation of dtype="object" for a column whose values are all numeric (and therefore the "soft coercion" to numeric type succeeds).

In this instance (in the absence of any non-numeric values in the series), even with my fix, it will still convert the dtype of the column from "object" to numeric.

I tried removing the conversion altogether.  That fixes #14849, but breaks another test (test_time_field_bug / #11324), so I will not commit that change.

I still think it is worth contributing my fix, since it improves upon the current situation.  A series with string values will not be converted to NaNs when doing groupby/apply on a DF with one or more datetime columns.

I will address other review feedback (re: "pd" import) and update the PR.

**Comment by jreback** [comment]
AFAICT this *only* resolved #15421 (w/o breaking anything else). so thats good.

**Comment by jreback** [comment]
#14423 I think *should* be fixable as well. want to give that a shot? (the example is not very good though).

File: pandas/tests/groupby/test_groupby.py (hunk 2 above)
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: add a 1-liner explaining this (maybe also make the test name a bit more descriptive)

**Comment by gwpdt** [comment]
Agreed - #15670 #15421 and #14423 are all fixed.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. merging shortly.

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**